### PR TITLE
New Recipe: basiclu v2.1.0

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

**Comment by amontoison** [comment]
@giordano  Do you have an idea how to fix the issue with `clock_gettime` routine ?

**Comment by giordano** [comment]
When else fails, read the source code:
```console
sandbox:${WORKSPACE}/srcdir/basiclu # grep -r -B2 '#define CLOCK_MONOTONIC_RAW\b' /opt/${target}/${target}
/opt/aarch64-apple-darwin20/aarch64-apple-darwin20/sys-root/usr/include/time.h-#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
/opt/aarch64-apple-darwin20/aarch64-apple-darwin20/sys-root/usr/include/time.h-_CLOCK_MONOTONIC_RAW __CLOCK_AVAILABILITY = 4,
/opt/aarch64-apple-darwin20/aarch64-apple-darwin20/sys-root/usr/include/time.h:#define CLOCK_MONOTONIC_RAW _CLOCK_MONOTONIC_RAW
```
At least for macOS, the `CLOCK_MONOTONIC_RAW` macro is defined when `_POSIX_C_SOURCE` is not defined (not a good case to deal with) or when `_DARWIN_C_SOURCE` is defined, which can be achieved with the `-D_DARWIN_C_SOURCE` flag to the compiler.  You'd have to check when the macro is defined for the other platforms as well.

File: B/basiclu/bundled/patches/cc.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
+--- a/basiclu/config.mk
```
Comment (giordano) [inline B/basiclu/bundled/patches/cc.patch:1]: Do you have to patch `config.mk`?  Can't you do
```
make CC99="cc -std=c99" LDLIBS="-lm -lrt"
```
?

Comment (amontoison) [reply B/basiclu/bundled/patches/cc.patch:1]: I don't need to patch `config.mk`, I can do 
```
make CC99="cc -std=c99" LDLIBS="-lm -lrt"
```
directly. I just forgot that I can do it. I will remove the patch.

File: B/basiclu/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    done
+fi
+cd basiclu/
+if [[ ${target} == *mingw32* ]]; then
+    make CC99="cc -std=c99"
+elif [[ ${target} == *-apple-* ]] || [[ ${target} == *freebsd* ]]; then
+    make CC99="cc -std=c99" -D_DARWIN_C_SOURCE
+else
+    make CC99="cc -std=c99" LDLIBS="-lm -lrt"
+fi
```
Comment (giordano) [inline B/basiclu/build_tarballs.jl]: I'd simplify to
```suggestion
if [[ "${target}" == *-apple-* ]] || [[ "${target}" == *freebsd* ]]; then
    CFLAGS="-D_DARWIN_C_SOURCE"
elif [[ "${target}" == *-linux-* ]]; then
    LDLIBS="-lm -lrt"
fi
make CC99="cc -std=c99" CFLAGS="${CFLAGS}" LDLIBS="${LDLIBS}"
```
but I'm not sure `-D_DARWIN_C_SOURCE` with FreeBSD is a good idea

Comment (amontoison) [reply B/basiclu/build_tarballs.jl]: Thanks !
For FreeBSD,
```
grep -r -B2 '#define CLOCK_MONOTONIC_RAW\b' /opt/${target}/${target}
```
returns nothing :(

Comment (amontoison) [reply B/basiclu/build_tarballs.jl]: Conclusion : Let's remove FreeBSD platform.

Comment (giordano) [reply B/basiclu/build_tarballs.jl]: Right, so maybe FreeBSD isn't really supported and you can skip it.  One option could be to define `CLOCK_MONOTONIC_RAW` as `CLOCK_MONOTONIC` (`-DCLOCK_MONOTONIC_RAW=CLOCK_MONOTONIC`), but I don't know what the consequences would be, I have no clue what this stuff does.

File: B/basiclu/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    UNAME="$(uname)"
+fi
+cd basiclu/
+if [[ "${target}" == *-apple-* ]]; then
+    CFLAGS="-D_DARWIN_C_SOURCE"
+elif [[ "${target}" == *-linux-* ]]; then
+    LDLIBS="-lm -lrt"
+fi
+if [[ "${target}" == *freebsd* ]]; then
+    make CC99="cc -std=c99" CFLAGS="${CFLAGS}" LDLIBS="${LDLIBS}" UNAME="${UNAME}" -DCLOCK_MONOTONIC_RAW=CLOCK_MONOTONIC
```
Comment (giordano) [inline B/basiclu/build_tarballs.jl:32]: `-D` is a compiler flag to set macros on the command line. Make doesn't know anything about that, you can't pass it that way. This Makefile accepts compiler flags with the variable `CFLAGS` so you can do the same as macOS above

File: B/basiclu/build_tarballs.jl (hunk 2 above)
Comment (amontoison) [reply B/basiclu/build_tarballs.jl]: If didn't work, I remove FreeBSD platform :)

**Comment by amontoison** [comment]
@giordano I don't know how to solve the issue with Windows platforms.
I can only build the shared library with the trick
```
cc -fPIC -shared -Wl,$(flagon --whole-archive) libbasiclu.lib -Wl,$(flagon --no-whole-archive) -o $libdir/libbasiclu.${dlext}
```
I will do a patch if it's the only solution.

**Comment by giordano** [comment]
```
[19:55:55] cc -std=c99  -DBASICLU_NOTIMER  -O3 -DNDEBUG -fPIC  build/basiclu_factorize.o build/basiclu_get_factors.o build/basiclu_initialize.o build/basiclu_object.o build/basiclu_solve_dense.o build/basiclu_solve_for_update.o build/basiclu_solve_sparse.o build/basiclu_update.o build/lu_build_factors.o build/lu_condest.o build/lu_dfs.o build/lu_factorize_bump.o build/lu_file.o build/lu_garbage_perm.o build/lu_initialize.o build/lu_internal.o build/lu_markowitz.o build/lu_matrix_norm.o build/lu_pivot.o build/lu_residual_test.o build/lu_setup_bump.o build/lu_singletons.o build/lu_solve_dense.o build/lu_solve_for_update.o build/lu_solve_sparse.o build/lu_solve_symbolic.o build/lu_solve_triangular.o build/lu_timer.o build/lu_update.o -o lib/libbasiclu.2.1.0.dll -lm
[19:55:55] /opt/i686-w64-mingw32/bin/../i686-w64-mingw32/sys-root/lib/../lib/libmingw32.a(lib32_libmingw32_a-crt0_c.o): In function `main':
[19:55:55] /workspace/srcdir/mingw-w64-v7.0.0/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain@16'
```
I'm not sure how they expect to generate a shared library without `-shared`

**Comment by amontoison** [comment]
> ```
> [19:55:55] cc -std=c99  -DBASICLU_NOTIMER  -O3 -DNDEBUG -fPIC  build/basiclu_factorize.o build/basiclu_get_factors.o build/basiclu_initialize.o build/basiclu_object.o build/basiclu_solve_dense.o build/basiclu_solve_for_update.o build/basiclu_solve_sparse.o build/basiclu_update.o build/lu_build_factors.o build/lu_condest.o build/lu_dfs.o build/lu_factorize_bump.o build/lu_file.o build/lu_garbage_perm.o build/lu_initialize.o build/lu_internal.o build/lu_markowitz.o build/lu_matrix_norm.o build/lu_pivot.o build/lu_residual_test.o build/lu_setup_bump.o build/lu_singletons.o build/lu_solve_dense.o build/lu_solve_for_update.o build/lu_solve_sparse.o build/lu_solve_symbolic.o build/lu_solve_triangular.o build/lu_timer.o build/lu_update.o -o lib/libbasiclu.2.1.0.dll -lm
> [19:55:55] /opt/i686-w64-mingw32/bin/../i686-w64-mingw32/sys-root/lib/../lib/libmingw32.a(lib32_libmingw32_a-crt0_c.o): In function `main':
> [19:55:55] /workspace/srcdir/mingw-w64-v7.0.0/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain@16'
> ```
> 
> I'm not sure how they expect to generate a shared library without `-shared`

Indeed, it's complicated without `-shared` ...
Thank you again for the help Mosè !

**Pull request merged**